ejb/WFLY-8449_Return_context_data_to_client.adoc

```diff
@@ -0,0 +1,81 @@
+= Return Context Data to EJB Clients
+:author:            Stuart Douglas
+:email:             sdouglas@redhat.com
+:toc:               left
+:icons:             font
+:keywords:          comma,separated,tags
+:idprefix:
+:idseparator:       -
+
+== Overview
+
+This will allow for some data that has been placed in the InvocationContext by an EJB or Interceptor to
+be returned to the client so it can be processed by client interceptors. Only data that has explicitly
+been requested by the interceptor will be returned.
+
+== Issue Metadata
+
+=== Issue
+
+* https://issues.jboss.org/browse/WFLY-8449[WFLY-8449]
+
+=== Related Issues
+
+* https://issues.jboss.org/browse/EAP7-976[EAP7-976]
+
+=== Dev Contacts
+
+* mailto:{email}[{author}]
+
+=== QE Contacts
+
+=== Affected Projects or Components
+
+ * Wildfly
+ * EJB Client
+
+=== Other Interested Projects
+
+== Requirements
+
+=== Hard Requirements
+
+An `EJBClientInterceptor` will be able to request data that it is interested in by calling
+`org.jboss.ejb.client.EJBClientInvocationContext#addReturnedContextDataKey(String key)`. This tells the server that the
+client interceptor would like any data that is stored under the provided key in the context data map to be returned
+to the client.
+
+Internally this will create/add to a `Set` in the clients context data stored under
+`org.jboss.ejb.client.invocation.returned-context-data`, which will then be sent to the server as part of the invocations
+context data, so no protocol change is required.
+
+If the server receives this set of keys then after the invocation has completed it will iterate the set of keys,
+and if a corresponding key name is found in the context data of the invocation then it will be copied into the
+context data map that is returned to the remote client. Any EJBClientInterceptor will be able to examine this data
+in the invocations context data map.
+
+This approach allows for a very fine grained approach to returning data to the client, so that only data that is actually
+required is returned.
+
+=== Nice-to-Have Requirements
+
+=== Non-Requirements
+
+It is not a requirement that all data in the context data map be returned to the client. This approach would
+be problematic for several reasons:
+
+ *  Most invocations simply do not want or need this data, so for the majority of clients this would increase network
+ and CPU usage for no reason
```

Comment:
This will cause a disparity with local remote invocation.  Maybe we should consider then dropping context data on local remote invocations as well unless the key is in the set?

Reply:
Thinking more about this, I think we _must_ make this change.  All the problems which are described with regards to serialization also apply with cloning; it's just less likely that a class might be missing in this case.

Reply:
I am pretty sure this data is not cloned on local invocations, at least it did not appear to be when I was looking at the code, I will double check tomorrow.